signature/src/mechanism/cosign/README.md

```diff
@@ -0,0 +1,207 @@
+# Cosign
+
+Cosign is short for Container Signing, Verification and Storage in an OCI registry.
+It aims to make signatures invisible infrastructure.
+[Cosign](https://github.com/sigstore/cosign/blob/main/README.md).
+
+## Policy Format
+
+Cosign is verified due to the container's policy configuration file.
+
+A Policy Requirement of Cosign should be like this
+
+```json
+{
+    "type": "sigstoreSigned",
+    "keyData": "<PUBKEY-DATA-IN-BASE64>",
+    "keyPath": "<URL-TO-THE-PUBKEY>",
+    "signedIdentity": <JSON-OBJECT>,
+},
+```
+
+Here, 
+* The `type` field must be `sigstoreSigned`, showing that this image is signed by `cosign`.
+* The `keyData` field includes the pubkey's content in base64.
+* The `keyPath` field indicates the pubkey's URL.
+* `signedIdentity` includes a JSON object, refer to [signedIdentity](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md#sigstoreSigned) for details.
+Because of the mechanism of Cosign, only `matchRepository` and `exactRepository`
+can be used to accept an image.
+
+**WARNING**: Must specify either `keyData` or `keyPath`, and must not both.
+
+## Implementation
+
+We wrap the [rust implementation](https://github.com/sigstore/sigstore-rs) for sigstore to fit
+in our defined interface for a signing scheme.
+
+## Work Flow
+
+Let's quickly see how Cosign works. For example, there's an image
+to be signed, named `172.17.0.3:5000/alpine:latest`.
+So as a reference, its
+* `registry`: `172.17.0.3:5000`, which is a test registry.
+* `repository`: `alpine`
+* `tag`: `latest`
+
+### Signing
+
+There will be three steps for the signing:
+- Signing the `Payload`
+- Wrap the `Payload` and signature into an image bundle
+- Upload the image bundle to the registry
+
+#### Signature Generate
+
+When we sign this image, we use a cosign-generated key pair.
+The digital signature algorithm is by default ECDSA-P256.
```

Comment:
ooi, how would I check that and how would I know if the default had been changed somehow?

Hmm, I've just found this, from https://github.com/sigstore/cosign/blob/main/specs/SIGNATURE_SPEC.md#signature-schemes:

> Implementations must support at least the following schemes:
> 
> - ECDSA-P256
> 
> No information about the signature scheme is included in the object. Clients must determine the signature scheme out-of-band during the verification process.

Why is the signature scheme not included though? Does that imply the verifier needs to attempt to verify against all the schemes it supports. If it fails to verify, it moves to the next scheme, _etc_? I might be missing something obvious, but if this is the way it works, it doesn't sound very efficient.

Reply:
AFAIK, now cosign only uses `ECDSA-P256` to sign and verify (from the source code), and it does not provide any signature scheme option.

Reply:
Sure!